### PR TITLE
[FIRRTL] Add a pass to specialize instance choices

### PR DESCRIPTION
**Pull request by nandor**
The `SpecializeOptions` pass eliminates instance choices and replaces them with instances targeting the modules associated with a case provided through the options.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 1]
```diff
@@ -850,4 +850,24 @@ def Lint :
...
+    If the design has option groups, instead of preserving the information
+    up to the emitted SystemVerilog, this pass specializes it early on for a
+    given choice.
+  }];
+  let constructor = "circt::firrtl::createSpecializeOptionPass()";
+
+  let options = [
+    Option<"selectedOption", "option", "std::string", "", "Option to specialize">,
+    Option<"selectedCase", "case", "std::string", "", "Case to pick">,
+  ];
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Passes.td:866]: What about multiple options? E.g., how would one specialize the test to be "Platform=FPGA" _and_ "Performance=Fast"?

Is this multiple options to the pass or is this handled by dynamically scheduling multiple passes to resolve things iteratively?

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 2]
```diff
@@ -197,6 +197,19 @@ def InstanceChoiceOp : HardwareDeclOp<"instance_choice", [
...
+      return getModuleNamesAttr()[0].cast<FlatSymbolRefAttr>();
+    }
+
+    /// Return the name of the option group.
+    StringAttr getOptionNameAttr() {
+      return getCaseNames()[0].cast<SymbolRefAttr>().getRootReference();
+    }
+
+    /// Return the target for an option or the default if none specified.
+    FlatSymbolRefAttr getTargetOrDefaultAttr(OptionCaseOp option);
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:212]: This can be landed separately from the actual pass.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                inst.getPortDirectionsAttr(), inst.getPortNamesAttr(),
+                inst.getAnnotationsAttr(), inst.getPortAnnotationsAttr(),
+                UnitAttr{}, inst.getInnerSymAttr());
+            inst.replaceAllUsesWith(newInst);
+            inst.erase();
+          });
+        });
+  }
+};
+} // namespace
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:61]: This pass defined statistics. However, they don't seem to be updated.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    if (!optionGroup) {
+      markAllAnalysesPreserved();
+      return;
+    }
+
+    auto optionCase = optionGroup.lookupSymbol<OptionCaseOp>(selectedCase);
+    if (!optionCase) {
+      markAllAnalysesPreserved();
+      return;
+    }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:40]: I don't think this is a great API---if these fail, then there's a user error and things arguably shouldn't proceed. Specifically, for something like the user specifies an option/case that doesn't exist or misspells it. If the pass silently does a no-op, then it's difficult to debug as it looks like the pass did the wrong thing.

tl;dr: more `signalPassFailure()` and tests of this behavior. 😉

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 5]
```diff
@@ -850,4 +850,24 @@ def Lint :
   let constructor = "circt::firrtl::createLintingPass()";
 }
 
+def SpecializeOption :
+    Pass<"firrtl-specialize-option", "firrtl::CircuitOp"> {
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Passes.td:854]: This could be a `FModuleOp` pass by my reading of https://mlir.llvm.org/docs/PassManagement/#operation-pass. What does the proposed error handling look like in that case?

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            if (optionGroup.getSymNameAttr() != inst.getOptionNameAttr())
+              return;
+
+            ImplicitLocOpBuilder builder(inst.getLoc(), inst);
+            auto newInst = builder.create<InstanceOp>(
+                inst->getResultTypes(), inst.getTargetOrDefaultAttr(optionCase),
+                inst.getNameAttr(), inst.getNameKindAttr(),
+                inst.getPortDirectionsAttr(), inst.getPortNamesAttr(),
+                inst.getAnnotationsAttr(), inst.getPortAnnotationsAttr(),
+                UnitAttr{}, inst.getInnerSymAttr());
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:54]: Any reason why the lower-to-bind isn't also copied over?

File: include/circt/Dialect/FIRRTL/Passes.td (hunk 1 above)
Comment (nandor) [reply include/circt/Dialect/FIRRTL/Passes.td:866]: In the current setup I'd schedule the pass multiple times. In a subsequent PR, when needed, I'd consider compacting the arguments and turning them into an array, but I think that's a bit clumsy with MLIR at the moment.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 6 above)
Comment (nandor) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:54]: The instance choice op doesn't have it.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 4 above)
Comment (nandor) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:40]: We might set up a flow for all designs and ask for specialization for a target, but a smaller input might have no use of options at all, in which case the pass would fail. I think it is reasonable to turn this into a no-op if there is no option to specialize.

File: include/circt/Dialect/FIRRTL/Passes.td (hunk 5 above)
Comment (nandor) [reply include/circt/Dialect/FIRRTL/Passes.td:854]: We'd get a bazillion copies of the error message? I think it's clearer handle parallelism within the pass in this case.

File: include/circt/Dialect/FIRRTL/Passes.td (hunk 1 above)
Comment (nandor) [reply include/circt/Dialect/FIRRTL/Passes.td]: The pass takes a list now.

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 7]
```diff
@@ -850,4 +850,24 @@ def Lint :
...
+    If the design has option groups, instead of preserving the information
+    up to the emitted SystemVerilog, this pass specializes it early on for a
+    given choice.
+  }];
+  let constructor = "circt::firrtl::createSpecializeOptionPass()";
+
+  let options = [
+    ListOption<"select", "select", "std::string", 
+               "Options to specialize, in option=case format">,
+  ];
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Passes.td:866]: This looks good!

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      }
+
+      std::string caseName = optionAndCase.substr(eq + 1);
+      auto caseOp = optionOp.lookupSymbol<OptionCaseOp>(caseName);
+      if (!caseOp) {
+        mlir::emitWarning(circuit.getLoc(), "invalid option case ") << caseName;
+        continue;
+      }
+      selected[StringAttr::get(&getContext(), optionName)] = caseOp;
+    }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:53]: Add tests for these, and for the error behavior that can result if `failed=true`.

File: test/Dialect/FIRRTL/specialize-option.mlir [hunk 9]
```diff
@@ -0,0 +1,29 @@
+// RUN: circt-opt --pass-pipeline='builtin.module(firrtl.circuit(firrtl-specialize-option{select=Platform=FPGA}))' %s | FileCheck %s
```
Comment (seldridge) [inline test/Dialect/FIRRTL/specialize-option.mlir:1]: Checking that the specialization of two options works would be good. This can be done with a second `RUN:` line and re-use the same test.

**Review by seldridge — APPROVED** [review]
Looks good.

Please add tests of all error/warning paths that this pass can take.
